### ISSUE TITLE
Join/group event messages together

### ISSUE THREAD

**Issue by Sherlouk** [rnystrom/GitHawk#1321]
I recall this being mentioned ages ago but we couldn't do it because the way (for example) added/removed label messages were constructed - But now they're just attributed labels we could start grouping them similar to GH.com

"BasThomas added `project-mgmt` `starter-task` and removed `accessibility` labels 18 hours ago"

**Comment by rnystrom** [comment]
This is a great starter if someone wants to peel it off

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by SD10** [comment]
https://github.com/rnystrom/GitHawk/issues/1541#issuecomment-366782371
> Closing bc I don’t want to reorder the timeline stuff on the client (complex and error prone).

Looked into this last night and I think it is a nightmare. It doesn't require an reordering per say, but rather a compression of timeline nodes.

Essentially, I need to turn an array of nodes such as:

`[A, B, B, B, C, B, B, D]` into `[A, BBB, C, BB, D]`
Where the B's are label added / removed events. 

Not impossible but it either requires double looping over nodes or maintaining a state of the last seen node while iterating. Definitely a little messy but I'll keep playing

**Comment by Sherlouk** [comment]
I took a quick look into this, it's super crude and will need to be made more efficient - but this is something that does the compression we need:

https://drive.google.com/open?id=1z_EpsXM2nUBVuE45lZIXyAM6IFc0n3ct

**Comment by SD10** [comment]
Stealing my work @Sherlouk 😂 Very similar to what I did but I used a `for loop`. 

The reason I chose a loop is because `previousResults` is immutable inside of the closure body. This requires a copy of the array to be made for every iteration. It get's expensive quickly for a `Collection` type. But we can use the `inout` version of reduce, `reduce(into:)` to get around this.

**Comment by SD10** [comment]
There would probably need to be two checks:
1. If last type seen is B
2. If last type seen is Z
Where Z is the new type that we're compressing B into

The reason is we need all the information inside of the unique B values so that we can resolve the final state of the label.

`foo added`, `foo removed`, `foo added` would result in `foo added` 😐 

That is where it kinda gets messy but IMO this processing should take part outside of the initial request/transformation.

Other things to consider, how do the time affect the labels?

SD10 added `foo`, `bar`, and removed `baz` 2 hours ago.

What happens if I come back 3 days later and then modify the labels more? Are they time sensitive?

SD10 added `foo`, `bar`, and removed `baz` 2 hours ago.
SD10 removed `foo`, `bar`, and added `baz` just now.

**Comment by SD10** [comment]
Update: Looks like the GitHub engineers don't care as much as I do 😓

<img width="572" alt="screen shot 2018-02-21 at 10 17 49 pm" src="https://user-images.githubusercontent.com/7445580/36520198-431bf7d0-1755-11e8-9700-e6bd88386ed8.png">

<img width="326" alt="screen shot 2018-02-21 at 10 15 31 pm" src="https://user-images.githubusercontent.com/7445580/36520192-372796aa-1755-11e8-9ef9-3cdd8ed9a07e.png">

**Comment by Sherlouk** [comment]
Yeah I'd probably back to the GitHub engineers on that one, don't think the time labels changed is super important!

Curious of other peoples opinions though, if a label was added and then removed it was likely a mistake - Should we just hide these? Keep the app a lot more streamlined of the net changes rather than each individual step?
